### ISSUE TITLE
Installation issue

### ISSUE THREAD

**Issue by Sebastien-Raguideau** [jakobnissen/vamb#16]
Hello,
Just starting and trying to install. When trying to import vamb, i get 
libpython3.7m.so.1.0: cannot open shared object file: No such file or directory
Does it mean I need python3.7?

**Comment by Sebastien-Raguideau** [comment]
I just created an environment with conda : 
conda create -n vamb python=3.7 PyTorch pysam Numpy
Upon trying to import vamb, I now get : 
ImportError: /home/sebr/seb/Applications/miniconda3/envs/vamb/lib/python3.7/site-packages/torch/lib/libtorch.so.1: undefined symbol: nvrtcGetProgramLogSize

**Comment by Sebastien-Raguideau** [comment]
Issue was solved by specifying PyTorch version 0.4.1. 
So 
conda create -n vamb python=3.7 PyTorch=0.4.1 pysam Numpy
works.

**Comment by jakobnissen** [comment]
Dear @Sebastien-Raguideau 

I just saw the issue now.
Vamb does not require Python 3.7 - but since the binary has been compiled to Python 3.7, you will need 3.7 to use the existing binary. Vamb runs with Python >= 3.5.
This is awkward, I know, so on the `dev` branch, we have made Vamb installable via pip, which automatically compiles the binaries upon installation (thanks to C. Titus Brown for the PR implementing that). The dev branch have significant changes, probably enough to warrant a Vamb 2.0, and will eventually be merged into `master`.

I've never seen the PyTorch-related problem before. Based on a little googling, it appears to be a PyTorch-internal issue (see e.g. [this issue](https://github.com/pytorch/pytorch/issues/10234))

**Comment by Sebastien-Raguideau** [comment]
Thanks for the answer. No issue anymore on my side, somehow just rolling back to older version of pytorch did the trick. 

So do you advise to look at the dev instead? 
 
Though I may have another question for you. I launched Vamb and it looks like it will never end. 
My dataset is about 823029 contigs over 49 samples. I edited run.py to change DEFAULT_PROCESSES from 8 to 90 as it looks like the variable for cpu parallelisation. I don't have access to gpu but I do have cpus.
And so things have been running since Thursday, cpu usage stay relatively low (~20) and we're now at Epoch12. 
Is this a technical issue or actually expected due to : the size of the dataset/ no cuda ?

**Comment by jakobnissen** [comment]
Yeah, `dev` is better in most aspects - however, I regularly commit to it, so there's a higher risk that any particular `dev` commit contains some horrific bug. Still, I'd run `dev`

Vamb can take a long time, but I'm still interested in optimizing its run time, so if you're seeing something unexpected, please tell me.

There are three phases which are time consuming:
1) BAM file parsing (see #15 ). This can be slow for enormous BAM files, or hundreds of BAM files, but shouldn't take days. It should be about 1.5x slower than `samtools view x.bam > /dev/null` run in parallel. See the issue for more in-depth discussion about it's speed.

2) Encoding. This is usually the most time-consuming part. On our cluster, using 8 CPU cores, it took about 3 hours for 180k sequences, so it should take about 1 day for your data.

3) Clustering. This scales O(N^2) unlike the other two steps (which are O(N)), so for very large datasets, this can dominate the time spent. However, for less than a million contigs, it should just take a few hours.

You can check the log file for the amount of time each step took. If you're curious if the program stopped working, you can see how long ago it last updated the logfile was.

**Comment by Sebastien-Raguideau** [comment]
Sorry my mistake, I said it has been running since Thursday, it has been since Tuesday so around ~3 days which is why I though it to be a problem. 
Looking at the log it is really the Encoding taking way too long.

**Comment by jakobnissen** [comment]
Alright, let's see what could cause this. Here are some things to check:

- Try to load in the `rpkm.npz` and `tnf.npz` files in Python (it's Numpy npz files). Do they have the expected dimensions of (823029 x 49) and (823029 x 136), respectively?

- Does PyTorch otherwise have normal speed when you run it on your computer or is it unexpectedly slow? Perhaps you can compare two toy networks on a laptop vs your cluster.

It could be possible that PyTorch can't handle 90 CPUs, though that does sound a bit weird.

**Comment by Sebastien-Raguideau** [comment]
I just looked at the matrices they do have the expected dimensions. 
I am not using PyTorch and can't really answer that question. 
Good idea looking at a small toy network could help. Would you have anything like that I could try?
